e2e_tests/tests/cluster/test_ship_logs.py

```diff
@@ -0,0 +1,524 @@
+            env = {
+                "DET_MASTER": "http://notreal.faketld",
+                "DET_SESSION_TOKEN": "token",
+                "DET_TASK_LOGGING_METADATA": "{}",
+                "http_proxy": srv.master_url(),
+            }
+            cmd = mkcmd("print('hello world')")
+            fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+            subprocess.run(fullcmd, env=env, check=True)
+            assert srv.logs == ["hello world\n"], srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_honor_no_proxy(self) -> None:
+        # Use a subprocess to control the environment.
+        with ShipLogServer() as srv:
+            env = {
+                "DET_MASTER": srv.master_url(),
+                "DET_SESSION_TOKEN": "token",
+                "DET_TASK_LOGGING_METADATA": "{}",
+                "http_proxy": "http://notreal.faketld",
+                "NO_PROXY": "127.0.0.1",
+            }
+            cmd = mkcmd("print('hello world')")
+            fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+            subprocess.run(fullcmd, env=env, check=True)
+            assert srv.logs == ["hello world\n"], srv.logs
+
+    @pytest.mark.e2e_cpu
+    def test_escape_hatch(self) -> None:
+        # Create a temporary directory to catch our escape-hatch logs
+        tmp = tempfile.mkdtemp(suffix="ship_logs")
+        try:
+            with ShipLogServer() as srv:
+                # Use a subprocess to control the environment.
+                # Leave out DET_MASTER to force a crash.
+                env = {"DET_SHIP_LOGS_PATH": tmp}
+                cmd = mkcmd("pass")
+                fullcmd = [sys.executable, "-u", ship_logs.__file__] + cmd
+                p = subprocess.run(fullcmd, env=env)
+                assert p.returncode == 80, p.returncode
+                assert srv.logs == [], srv.logs
+                files = os.listdir(tmp)
+                assert len(files) == 2, files
+                assert "ship-logs-ran" in files, files
+                files.remove("ship-logs-ran")
+                with open(os.path.join(tmp, files[0]), "r") as f:
+                    text = f.read()
+                assert "KeyError: 'DET_MASTER'" in text, text
+        finally:
+            shutil.rmtree(tmp)
+
+
+class TestReadNewlinesOrCarriageReturns:
+    # read_newlines_or_carriage_returns is designed to read from filedescriptors resulting from
+    # subprocess.Popen(bufsize=0).stdout, and different kinds of filehandles can result in slightly
+    # different read/write behaviors, and stdout and stderr can additionally be different on
+    # different operating systems, so this test will use Popen to create file descriptors instead of
+    # something more convenient like os.pipe(), in order to test against the most realistic
+    # conditions.
+
+    @pytest.mark.e2e_cpu
+    @pytest.mark.parametrize("lastline", [repr("3"), repr("3\r"), repr("3\n")])
+    def test_process_killed(self, lastline: str) -> None:
```

Comment:
I think this test actually tests that `read_newlines_or_carriage_returns` returns as soon as its IO is terminated by whatever happens when a process is killed. But I don't know how we could write `read_newlines...` in any other way.

Ohhh... there's subtle stuff going on with the "." character. The proc isn't killed until _after_ that character is printed to the stream. But i's hard for me to work out what behavior the test is supposed to test.

My recommendation is to:
* Remove all the assertions that don't test the specific edge case this test is supposed to test. If there's more than one, then that recommendation extends to replicating the test with different assertions.
* Change the test's name to `test_read_newlines_...` for the thing being tested
* Give the remaining assertions an `expression2` that a reader/debugger can understand the intent of at a glance

Reply:
The behavior I'm trying to test is exactly what I said... make sure the reader behaves well when reading from the stdout of a process that is killed.

I think you're right that the behavior being observed is "whatever happens when a process is killed".  That's what I want to test.

The only other way I know of for a file descriptor like this to break is if the process owning the write side of the pipe closes it, which is what the other test is for.

Ultimately, what I wanted to prove was: "do I need an exception handler around fd.read() or not?".  The answer seems to be no.

- I'm happy to add more useful messages to the asserts.
- Can you elaborate more on "it's hard for me to work out what behavior the test is supposed to test"?
- For "Remove all the assertions that don't test the specific edge case this test is supposed to test", you mean the `print("1\n")` and `print("2\r")`?  Are there others you had in mind?

Reply:
You are absolutely right. I overlooked that and it was the wrong recommendation. The other parts we worked out over the phone.